src/jit/morph.cpp

```diff
@@ -10905,9 +10905,9 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
 
         if (fgGlobalMorph)
         {
-#if !FEATURE_STACK_FP_X87
+#if !FEATURE_STACK_FP_X87 || defined(LEGACY_BACKEND)
```

Comment:
As Bruce pointed out, this is always true. I suspect (though haven't done any historical analysis) that since the time the `LEGACY_BACKEND` was fully functional, either the condition in value numbering was tightened, or the IR was changed in some other way.
While it would be interesting to understand what changed, I think that it is always correct to add these casts, so doing this unconditionally seems like the right fix to me.